core/src/main/java/org/elasticsearch/script/ScriptType.java

```diff
@@ -24,66 +24,117 @@
 import org.elasticsearch.common.io.stream.StreamOutput;
 
 import java.io.IOException;
-import java.util.Locale;
+import java.util.Objects;
 
 /**
- * The type of a script, more specifically where it gets loaded from:
- * - provided dynamically at request time
- * - loaded from an index
- * - loaded from file
+ * ScriptType represents the way a script is stored and retrieved from the {@link ScriptService}.
+ * It's also used to by {@link ScriptSettings} and {@link ScriptModes} to determine whether or not
+ * a {@link Script} is allowed to be executed based on both default and user-defined settings.
  */
 public enum ScriptType {
 
-    INLINE(0, "inline", "inline", false),
-    STORED(1, "id", "stored", false),
-    FILE(2, "file", "file", true);
+    /**
+     * INLINE scripts are specified in numerous queries and compiled on-the-fly.
+     * They will be cached based on the lang and code of the script.
+     * They are turned off by default for security purposes.
+     */
+    INLINE ( 0 , "inline" , new ParseField("inline")       , false ),
 
-    private final int val;
-    private final ParseField parseField;
-    private final String scriptType;
-    private final boolean defaultScriptEnabled;
+    /**
+     * STORED scripts are saved as part of the {@link org.elasticsearch.cluster.ClusterState}
+     * based on user requests.  They will be cached when they are first used in a query.
+     * They are turned off by default for security purposes.
+     */
+    STORED ( 1 , "stored" , new ParseField("stored", "id") , false ),
+
+    /**
+     * FILE scripts are loaded from disk either on start-up or on-the-fly depending on
+     * user-defined settings.  They will be compiled and cached as soon as they are loaded
+     * from disk.  They are turned on by default as they should always be safe to execute.
+     */
+    FILE   ( 2 , "file"   , new ParseField("file")         , true  );
 
+    /**
+     * Reads an int from the input stream and converts it to a {@link ScriptType}.
+     * @return The ScriptType read from the stream. Throws an {@link IllegalStateException} if no ScriptType is found based on the id.
+     */
     public static ScriptType readFrom(StreamInput in) throws IOException {
-        int scriptTypeVal = in.readVInt();
-        for (ScriptType type : values()) {
-            if (type.val == scriptTypeVal) {
-                return type;
-            }
-        }
-        throw new IllegalArgumentException("Unexpected value read for ScriptType got [" + scriptTypeVal + "] expected one of ["
-                + INLINE.val + "," + FILE.val + "," + STORED.val + "]");
-    }
+        int id = in.readVInt();
 
-    public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOException{
-        if (scriptType != null) {
-            out.writeVInt(scriptType.val);
+        if (FILE.id == id) {
```

Comment:
This is a bit awkward because the id is not considered a compile-time constant since it's set through the enum constructor.  One option would be to add compile time constants for each id and use that instead, but I think the if statement is probably simpler.